File: backend/hackathon-backend/routers/file_upload.py
"""
File upload router for handling general file uploads
"""

from fastapi import APIRouter, UploadFile, File, HTTPException, Depends
from typing import List
import os

from services.file_upload_service import file_upload_service
from schemas import FileUploadResponse, PresignedUrlResponse

router = APIRouter()


@router.post("/upload", response_model=FileUploadResponse)
async def upload_file(
    file: UploadFile = File(...),
    prefix: str = ""
):
    """
    Upload a file to configured storage (S3/MinIO/Local)
    
    - **file**: The file to upload
    - **prefix**: Optional prefix for organizing files (e.g., "documents", "images")
    """
    try:
        upload_result = await file_upload_service.upload_file(file, prefix)
        return upload_result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"File upload failed: {str(e)}")


@router.post("/upload-multiple", response_model=List[FileUploadResponse])
async def upload_multiple_files(
    files: List[UploadFile] = File(...),
    prefix: str = ""
):
    """
    Upload multiple files to configured storage
    
    - **files**: List of files to upload
    - **prefix**: Optional prefix for organizing files
    """
    results = []
    
    for file in files:
        try:
            upload_result = await file_upload_service.upload_file(file, prefix)
            results.append(upload_result)
        except HTTPException as e:
            # Skip failed files but continue with others
            results.append({
                "filename": file.filename,
                "error": e.detail,
                "success": False
            })
        except Exception as e:
            results.append({
                "filename": file.filename,
                "error": str(e),
                "success": False
            })
    
    return results


@router.get("/presigned-url/{filename}", response_model=PresignedUrlResponse)
async def get_presigned_url(
    filename: str,
    expires_in: int = 3600
):
    """
    Generate a presigned URL for accessing a file
    
    - **filename**: The filename to generate URL for
    - **expires_in**: URL expiration time in seconds (default: 1 hour)
    """
    presigned_url = file_upload_service.get_presigned_url(filename, expires_in)
    
    if not presigned_url:
        raise HTTPException(status_code=404, detail="File not found or presigned URL generation failed")
    
    from datetime import datetime, timedelta
    expires_at = datetime.now() + timedelta(seconds=expires_in)
    
    return {
        "filename": filename,
        "presigned_url": presigned_url,
        "expires_at": expires_at
    }


@router.delete("/{filename}")
async def delete_file(filename: str):
    """
    Delete a file from storage
    
    - **filename**: The filename to delete
    """
    success = await file_upload_service.delete_file(filename)
    
    if not success:
        raise HTTPException(status_code=404, detail="File not found or deletion failed")
    
    return {"message": "File deleted successfully", "filename": filename}

File: backend/node-backend/src/routes/fileUpload.js
"use strict";

const express = require('express');
const multer = require('multer');
const fileUploadService = require('../services/fileUploadService');
const { authenticateToken } = require('../middleware/auth');

const router = express.Router();

// Configure multer for file uploads
const upload = multer({
    limits: {
        fileSize: 10 * 1024 * 1024, // 10MB limit
    },
    fileFilter: (req, file, cb) => {
        const allowedTypes = [
            'image/jpeg',
            'image/png', 
            'image/gif',
            'image/webp',
            'application/pdf'
        ];
        
        if (allowedTypes.includes(file.mimetype)) {
            cb(null, true);
        } else {
            cb(new Error('File type not allowed'), false);
        }
    }
});

// Upload single file
router.post('/upload', authenticateToken, upload.single('file'), async (req, res) => {
    try {
        const { prefix = '' } = req.body;
        
        if (!req.file) {
            return res.status(400).json({ 
                error: 'No file provided' 
            });
        }
        
        const uploadResult = await fileUploadService.uploadFile(req.file, prefix);
        res.status(200).json(uploadResult);
        
    } catch (error) {
        console.error('File upload error:', error);
        res.status(500).json({ 
            error: 'File upload failed', 
            details: error.message 
        });
    }
});

// Upload multiple files
router.post('/upload-multiple', authenticateToken, upload.array('files'), async (req, res) => {
    try {
        const { prefix = '' } = req.body;
        
        if (!req.files || req.files.length === 0) {
            return res.status(400).json({ 
                error: 'No files provided' 
            });
        }
        
        const results = [];
        
        for (const file of req.files) {
            try {
                const uploadResult = await fileUploadService.uploadFile(file, prefix);
                results.push({
                    ...uploadResult,
                    success: true
                });
            } catch (error) {
                results.push({
                    filename: file.originalname,
                    error: error.message,
                    success: false
                });
            }
        }
        
        res.status(200).json(results);
        
    } catch (error) {
        console.error('Multiple file upload error:', error);
        res.status(500).json({ 
            error: 'Multiple file upload failed', 
            details: error.message 
        });
    }
});

// Get presigned URL for file access
router.get('/presigned-url/:filename', authenticateToken, async (req, res) => {
    try {
        const { filename } = req.params;
        const { expires_in = 3600 } = req.query;
        
        const expiresIn = parseInt(expires_in);
        
        const presignedUrl = fileUploadService.getPresignedUrl(filename, expiresIn);
        
        if (!presignedUrl) {
            return res.status(404).json({ 
                error: 'File not found or presigned URL generation failed' 
            });
        }
        
        const expiresAt = new Date(Date.now() + expiresIn * 1000);
        
        res.status(200).json({
            filename,
            presigned_url: presignedUrl,
            expires_at: expiresAt.toISOString()
        });
        
    } catch (error) {
        console.error('Presigned URL error:', error);
        res.status(500).json({ 
            error: 'Failed to generate presigned URL', 
            details: error.message 
        });
    }
});

// Delete file
router.delete('/:filename', authenticateToken, async (req, res) => {
    try {
        const { filename } = req.params;
        
        const success = await fileUploadService.deleteFile(filename);
        
        if (!success) {
            return res.status(404).json({ 
                error: 'File not found or deletion failed' 
            });
        }
        
        res.status(200).json({ 
            message: 'File deleted successfully', 
            filename 
        });
        
    } catch (error) {
        console.error('File deletion error:', error);
        res.status(500).json({ 
            error: 'File deletion failed', 
            details: error.message 
        });
    }
});

module.exports = router;